File: challenge/strings/permutation/perm.ts
function registerLetters(
    str: string, 
    inventory: object = {}, 
    increment: number = 1
): object {

    for (let c of str) {
        inventory[c] = c in inventory ? inventory[c] + increment : increment;
    }
    return inventory;
}

function checkLetters(str: string, inventory: object): object {
    return registerLetters(str, inventory, -1);
}

function compareLetterInventories(s1: string, s2: string): object {
    return checkLetters(s2, registerLetters(s1));
}


function isPermutation(baseStr: string, compareStr: string): boolean {
    let comparison = compareLetterInventories(baseStr, compareStr);
    return Object.values(comparison).every(c => c == 0);
}

function test(s1: string, s2: string): void {
    console.log(`${s1} vs ${s2}: ${isPermutation(s1, s2)}`);
}

let testValues = [
    ["abba", "baab"],
    ["dad", "add"],
    ["bacon", "eggs"],
    ["shut", "tush"]
];

testValues.forEach(([a, b]) => test(a, b));

File: challenge/strings/permutation/perm.py
"""
Write an algorithm to determine if two strings are permutations of each other.
2024/05/07
"""

def is_permutation(base_string, compare_string):
    """ 
    Given two strings, return True if the second is a permutation of the
    first, otherwise False. It is a permutation if it contains exactly the
    same letters in the same quantities.

    We register the count of letter in the first string in a dictionary, then
    for the second string we do the same in reverse. If the second is a
    permutation of the first, the result will be all zero values.
    """
    
    def register_letters(string, inventory = {}, increment = 1):
        for c in string:
            inventory[c] = inventory[c] + increment if c in inventory else 1
        return inventory

    def check_letters(string, inventory):
        return register_letters(string, inventory, -1)

    def compare_letter_inventories(s1, s2):
        return check_letters(s2, register_letters(s1))

    comparison = compare_letter_inventories(base_string, compare_string)
    return all(c == 0 for c in comparison.values())

def test(s1, s2):
    print(f"{s1} vs {s2}: {is_permutation(s1, s2)}")

test_values = [
    ('abba', 'baab'),
    ('dad', 'add'),
    ('bacon', 'eggs'),
    ('shut', 'tush')
]

for (s1, s2) in test_values:
    test(s1, s2)










